README.md
## 制图模板管理

![生成制图模板](https://i.loli.net/2019/11/14/9a5rMFyg8RZxWlL.gif)

### 一、使用
#### 1.1 命令
```
安装依赖
$ npm install

生成地图模板
$ npm run test

生成jsdoc
$ npm run jsdoc
```

#### 1.2 选择你的地图版式
```
const A0WallMap = require('./template/A0-horizontal'); // 横板
const A0WallMap = require('./template/A0-vertical'); // 竖版
```

#### 1.3 实例化你的地图边框主题
```
在 `border_themes`文件夹下查找想要的边框主题。
文件名即函数参数

const A0 = new A0WallMap('A0verticalDefault');
```
|              主题            |          默认参数        |     是否为默认     |     类型     |
| ---------------------------- | ----------------------- | ------------------ | ----------- |
|          A0竖版蓝边框         |    A0VerticalDefault    |         是         |   String    |
|          A0横版蓝边框         |    A0HorizontalDefault  |         是         |   String    |

#### 1.4 地图元素
`npm run jsdoc` 生成注释文档

查看你需要引入的地图元素配置参数

地图元素函数：

|              名称            |            函数          | 参数 (详情查看jsdoc注释文档) |          默认参数        |     类型     |
| ---------------------------- | ----------------------- | -------------------------- | ------------------------ | ----------- |
|            地图               |      createMap()        |             是             |           -              |   String    |
|            指北针             |      createCompass()    |             否             |    './assets/zbz.png'    |   String    |
|             边框              |      createBorder()    |             否             |           -              |       -       |
|            地图标题           |      createTiltle()     |             是             |           -              |   String    |
|            图例               |      createLegend()     |             是            |           -              |    Object    |
|            主编单位           |      createZbdw()       |             否            |  国信司南（北京）地理信息技术有限公司 |   String    |
|            比例尺             |      createScale()      |             是            |           -              |   String    |
|            承编单位           |      createCbdw()       |             否            |  国信司南（北京）地理信息技术有限公司 |   String    |

![生成jsdoc文档](https://i.loli.net/2019/11/14/3NW45ISx2AtDPKQ.gif)

#### 1.5 生成模板
```
A0.output();
```


### 二、横竖版默认配置
A0竖版配置

|              名称            |    width    |   height  | borderLeftW | borderTopW | borderRightW | borderBottomW |
| ---------------------------- | ----------- | --------- | ----------- | ---------- | ---------- | ---------- |
|             canvas           |    9933     |   14041   |       -     |      -     |       -    |       -    |
|           默认地图            |    8986     |   12698   |       -     |       -    |       -    |       -    |
|  默认边框(A0VerticalDefault)  |    8986     |   12698   |     467     |     880    |     465    |      463   |

<br>

A0横板配置

|              名称            |    width    |   height  | borderLeftW | borderTopW | borderRightW | borderBottomW |
| ---------------------------- | ----------- | --------- | ----------- | ---------- | ---------- | ---------- |
|             canvas           |    14041     |   9933   |       -     |      -     |       -    |       -    |
|           默认地图            |    13111     |   8588   |       -     |       -    |       -    |       -    |
|  默认边框(A0VerticalDefault)  |    8986     |   12698   |     467     |     880    |     463    |      465   |

### 三、🌹 如果你有更好的边框主题
#### 3.1 在 `border_themes` 文件夹下建立你的边框主题文件夹。
* 文件夹名以 `尺寸 + 版式 + 样式` 命名 (例如：A0VerticalBlueLace);
* 建立 `index.js` , 引入你的边框图片;

#### 3.2 关于 `index.js` (格式以下为准)
```
'use strict';
// A0竖版板默认边框
module.exports = {
  borderLeftW: 467,
  borderTopW: 880,
  borderRightW: 465,
  borderBottomW: 463,
  borderImg: './boder_themes/A0VerticalDefault/border.png',
};
```

🌹 提供格式不对，我们在最初做了拦截，请遵守格式规定。

#### 3.3 A0边框主题

|              主题            |          默认参数         |     是否为默认     |     类型     |
| ---------------------------- | ------------------------ | ------------------ | ----------- |
|            A0竖版            |           /              |         /          |      /      |
|          A0竖版蓝边框         |    A0VerticalDefault     |         是         |   String    |
|          A0竖版蓝花边框       |    A0VerticalBlueLace    |         否         |   String    |
|             A0横版           |            /             |          /         |      /      |
|          A0横版蓝边框         |    A0HorizontalDefault   |         是         |   String    |
|          A0横版蓝花边框        |    A0HorizontalBlueLace  |         否         |   String    |
|          A0横版黑花边边框      |    A0HorizontalBlackLace  |         否         |   String    |

### 四、[版本](https://github.com/geocompass/MapPT/blob/master/CHANGELOG.md)

### 五、[未来计划](https://github.com/geocompass/MapPT/blob/master/PLANS.md)

CHANGELOG.md
## 2019-12-4
* 修改了比例尺与图示相对位置的同步改变

### change

* **默认值:** 添加了主题边框默认值
* **默认值:** 添加了指北针默认值
* **默认值:** 添加了图例边框默认值
* **添加:** 增加了所有必填参数检测拦截判断
* **添加:** 增加了制图过程进度

test.js
'use strict';
// 详细参数配置请生成jsdoc文档查阅
const A0WallMap = require('./template/A0-vertical'); // 版式 vertical Horizontal
(async function () {
  const A0 = new A0WallMap('A0VerticalDefault'); // 实例化主题
  await A0.createMap('./assets/merge.png'); // 地图(必填)
  await A0.createCompass('./assets/zbz.png'); // 指北针
  await A0.createBorder(); // 边框
  A0.createTiltle('地图模板标题'); // 标题(必填)
  await A0.createLegend({
    data: [{
      icon: './assets/qiren.png',
      txt: '图例',
    }],
    option: {
      iconColumn: 1,
      iconWidth: 176,
      iconHight: 176,
      iconFontSize: 105,
      iconFontColor: '#221714',
    },
  }, {
    boderColor: '#0A5179',
    tuliFontSize: 146,
    tuliFontColor: '#221815',
  }); // 图例(必填)
  A0.createZbdw(); // 主编单位
  A0.createScale('1500'); // 比例尺(必填)
  A0.createCbdw(); // 承编单位
  A0.output();
})();

template/A0-vertical.js
'use strict';
const {
  createCanvas,
  loadImage
} = require('canvas');
const BaseTemplate = require('./template-base.js');

/**
 * Class 竖图
 * @extends BaseTemplate
 */
class A0Vertical extends BaseTemplate {
  constructor(props = 'A0VerticalDefault') {
    super();
    this.canvas = new createCanvas(9933, 14041);
    this.ctx = this.canvas.getContext('2d');
    this.canvasW = 9933;
    this.canvasH = 14041;
    this.proxyProps = this.proxyProps(props);
    this.boderThemes = require(`../boder_themes/${props}`);
    this.proxyBorderThemes = this.proxyBorderThemes(this.boderThemes);
    this.borderLeftW = this.boderThemes.borderLeftW;
    this.borderTopW = this.boderThemes.borderTopW;
    this.borderRightW = this.boderThemes.borderRightW;
    this.borderBottomW = this.boderThemes.borderBottomW;
    this.borderImg = this.boderThemes.borderImg;
  }

  // proxy
  proxyProps(props) {
    if (!props) throw new RangeError('未选择边框主题');
  }
  proxyBorderThemes(boderThemes) {
    const boderItem = ['borderLeftW', 'borderTopW', 'borderRightW', 'borderBottomW', 'borderImg'];
    boderItem.forEach(el => {
      if (!(el in boderThemes)) {
        throw new RangeError('边框主题参数错误');
      }
    });
  }

  /**
   * 地图
   * @param {string} mapImage - 地图路径
   * @param {object} [paramObject={}] - 地图定位参数
   * @param {number} [paramObject.left=this.borderLeftW] - 地图相对左边距
   * @param {number} [paramObject.top=this.borderTopW] - 地图相对上边距
   */
  async createMap(mapImage, {
    left = this.borderLeftW,
    top = this.borderTopW
  } = {}) {
    if (!mapImage) throw new RangeError('地图路径参数错误');
    const {
      ctx
    } = this;
    const imageBuffer = await loadImage(mapImage);
    ctx.drawImage(imageBuffer, left, top);
    console.group('竖版图幅制作中▼');
    console.log('地图✔');
  }

  /**
   * 图幅指北针
   * @param {string} [compassPath='./assets/zbz.png'] - 图幅指北针路径
   * @param {object} [paramObject={}] - 图幅指北针定位参数
   * @param {number} [paramObject.distanceBorderRight=354] - 图幅指北针距右边框
   * @param {number} [paramObject.distanceBorderTop=360] - 图幅指北针距上边框
   * @param {number} [paramObject.imgW=708] - 图幅指北宽
   * @param {number} [paramObject.imgH=877] - 图幅指北针高
   */
  async createCompass(compassPath = './assets/zbz.png', {
    distanceBorderRight = 354,
    distanceBorderTop = 360,
    imgW = 708,
    imgH = 877
  } = {}) {
    const {
      ctx
    } = this;
    const [left, top] = [this.canvasW - imgW - distanceBorderRight - this.borderRightW, distanceBorderTop + this.borderTopW];
    const imageBuffer = await loadImage(compassPath);
    ctx.drawImage(imageBuffer, left, top, imgW, imgH);
    console.log('图幅指北针√');
  }

  /**
   * 图幅边框
   * @param {string} [borderPath=this.borderImg] - 图幅边框路径
   * @param {object} [paramObject={}] - 图幅边框定位参数
   * @param {number} [paramObject.left=0] - 图幅边框相对左边距
   * @param {number} [paramObject.top=0] - 图幅边框相对上边距
   */
  async createBorder(borderPath = this.borderImg, {
    left = 0,
    top = 0
  } = {}) {
    const {
      ctx
    } = this;
    const imageBuffer = await loadImage(borderPath);
    ctx.drawImage(imageBuffer, left, top);
    console.log('图幅边框√');
  }

  /**
   * 图幅标题
   * @param {string} title - 图幅字体名称
   * @param {object} [paramObject={}] - 图幅标题定位参数
   * @param {number} [paramObject.left=2921] - 图幅字体相对左边距
   * @param {number} [paramObject.top=319] - 图幅字体相对上边距
   * @param {number} [paramObject.fontSize=417] - 图幅字体字号
   * @param {string} [paramObject.fontColor='#0A5179'] - 图幅字体颜色
   * @param {string} [paramObject.fontType='hyf'] - 图幅字体
   */
  createTiltle(title, {
    left = 2921,
    top = 319,
    fontSize = 417,
    fontColor = '#0A5179',
    fontType = 'hyf'
  } = {}) {
    if (!title) throw new RangeError('图幅标题参数错误');
    const {
      ctx
    } = this;
    ctx.font = `${fontSize}px ${fontType}`;
    ctx.fillStyle = fontColor;
    ctx.textBaseline = 'top';
    ctx.fillText(title, left, top);
    console.log('图幅标题√');
  }

  /**
   * 图例
   * @param {object} iconOption - 图例指示内容
   * @param {Array.<{icon: String, txt: String}>} iconOption.data - 图例指示内容数组
   * @param {string} iconOption.data[].icon - 图例指示内容图片路径
   * @param {string} iconOption.data[].txt - 图例指示内容图片信息
   * @param {object} iconOption.option - 图例指示内容配置项
   * @param {number} iconOption.option.iconColumn - 图例内容需要分多少列(必须可以被iconOption.data.length整除)
   * @param {number} iconOption.option.iconWidth - icon宽
   * @param {number} iconOption.option.iconHight - icon高
   * @param {number} iconOption.option.iconFontSize - icon字体大小
   * @param {string} iconOption.option.iconFontColor - icon字体颜色
   * @param {object} LegendOption 图例配置项
   * @param {string} [LegendOption.boderColor='#0A5179'] - 图例边框颜色
   * @param {number} [LegendOption.tuliFontSize=146] - 图例标题字号
   * @param {string} [LegendOption.tuliFontColor='#221815'] - 图例标题颜色
   */
  async createLegend(
    iconOption = {},
    LegendOption = {
      boderColor: '#0A5179',
      tuliFontSize: 146,
      tuliFontColor: '#221815'
    }
  ) {
    if (!iconOption.data || !iconOption.option) throw new RangeError('图例参数错误');
    const {
      ctx
    } = this;
    const [boderColor, tuliFont, tuliColor, pointData, iconColumn, iconW, iconH, iconFont, iconColor] = [
      LegendOption.boderColor,
      LegendOption.tuliFontSize,
      LegendOption.tuliFontColor,
      iconOption.data,
      iconOption.option.iconColumn,
      iconOption.option.iconWidth,
      iconOption.option.iconHight,
      iconOption.option.iconFontSize,
      iconOption.option.iconFontColor,
    ];
    // 计算总高度
    const iconColumnNums = Math.ceil(pointData.length / iconColumn); // 个数/列
    const tuliMargin = tuliFont * (3.5); // 图例上边距(0.5) + 图例高(1) + 图例下边距(1) + icon距下边框(1)
    const iconColumnH = iconH * (2 * iconColumnNums - 1); // 纵列icon的总高度(icon纵列间距1)
    const borderH = tuliMargin + iconColumnH; // 图例边框总高度
    // 计算总宽度
    const countPointDataLen = [];
    pointData.forEach(item => {
      countPointDataLen.push(item.txt.length);
    });
    const iconTxtMaxLen = countPointDataLen.sort((a, b) => {
      return b - a;
    })[0] * iconFont; // 计算最长的txt
    const iconMaxLen = iconTxtMaxLen + iconW * (1.5); // icon中最大长度
    let borderW; // iconMaxLen * 个数 + icon间距(0.5) + 两边间距(0.5+0.5)
    if (iconMaxLen > 560) {
      borderW = iconColumn * iconMaxLen + (iconColumn - 1) * iconMaxLen / 2 + 200 + 200; // 最大宽度 = 最长icon * 个数 + 最长icon * 间距 + 200 + 200
    } else {
      if (iconColumn === 1) {
        borderW = 560 + 200 + 200; // 最大宽度 = 图例宽度 + 200 + 200
      } else {
        borderW = iconColumn * iconMaxLen + (iconColumn - 1) * iconMaxLen / 2 + 200 + 200;
      }
    }

    // 创建、定位自适应边框
    const [borderRight, borderBottom, borderLineW] = [this.borderRightW, this.borderBottomW, 10];
    const [borderLeft, borderTop] = [this.canvasW - borderRight - borderW - borderLineW, this.canvasH - borderBottom - borderH - borderLineW];
    ctx.strokeStyle = boderColor;
    ctx.fillRect(borderLeft, borderTop, borderW + 20, borderH + 20);
    ctx.clearRect(borderLeft + 10, borderTop + 10, borderW, borderH);
    ctx.fillStyle = '#ffffff';
    ctx.fillRect(borderLeft + 10, borderTop + 10, borderW, borderH);

    // 创建、定位自适应'图 例'
    const [tu, li] = ['图', '例'];
    const fontTuLeft = borderLeft + (borderW - 560) / 2;
    const fontLiLeft = fontTuLeft + tuliFont + 308;
    const tuLiTop = borderTop + 0.5 * tuliFont;
    ctx.font = `${tuliFont}px hanyizhongyuan`;
    ctx.fillStyle = tuliColor;
    ctx.fillText(tu, fontTuLeft, tuLiTop);
    ctx.fillText(li, fontLiLeft, tuLiTop);

    // 创建、定位自适应多列icon
    pointData.forEach(async (item, index) => {
      const key = index + 1;
      const iconFirstLine = index % iconColumnNums;
      const iconFirstColumn = Math.ceil(key / iconColumnNums);
      const iconIndexLeft = borderLeft + 200 + (iconFirstColumn - 1) * iconMaxLen + (iconFirstColumn - 1) * (iconMaxLen / 2); // borderLeft + 左边距 + 第几列的icon + 第几列的icon间距
      const iconIndexTop = tuLiTop + 2 * tuliFont + iconH * 2 * iconFirstLine;
      const imageBuffer = await loadImage(item.icon);
      ctx.drawImage(imageBuffer, iconIndexLeft, iconIndexTop, iconW, iconH);
      ctx.fillStyle = iconColor;
      ctx.font = `${iconFont}px hanyizhongyuan`;
      ctx.fillText(item.txt, iconIndexLeft + 1.5 * iconW, iconIndexTop + (iconH - iconFont) / 2);
    });
    console.log('制作图例√');
  }

  /**
   * 主编单位
   * @param {string} [zbdwName='国信司南（北京）地理信息技术有限公司'] - 主编单位字体名称
   * @param {object} [paramObject={}] - 主编单位定位参数
   * @param {number} [paramObject.left=713] - 主编单位字体相对左边距
   * @param {number} [paramObject.top=13803] - 主编单位字体相对上边距
   * @param {number} [paramObject.fontSize=104] - 主编单位字体字号
   * @param {string} [paramObject.fontColor='#221815'] - 主编单位字体颜色
   */
  createZbdw(zbdwName = '国信司南（北京）地理信息技术有限公司', {
    left = 713,
    top = 13803,
    fontSize = 104,
    fontColor = '#221815'
  } = {}) {
    const name = `主编单位：${zbdwName}`;
    const {
      ctx
    } = this;
    ctx.font = `${fontSize}px hanyizhongyuan`;
    ctx.fillStyle = fontColor;
    ctx.fillText(name, left, top);
    console.log('主编单位√');
  }

  /**
   * 比例尺数值转百分符
   * @param {string} scale 数值例如3000
   * @return {String} 转换百分符的比例尺数值例如3 000
   */
  percentSign(scale) {
    return scale.replace(/(\d)(?=(?:\d{3})+$)/g, '$1 ');
  }
  /**
   * 比例尺数值1：xxxx
   * @param {number} scale 比例尺数值
   * @param {object} [paramObject={}] - 比例尺数值定位参数
   * @param {number} [paramObject.left=4000] - 比例尺数值字体相对左边距
   * @param {number} [paramObject.top=13812] - 比例尺数值字体相对上边距
   * @param {number} [paramObject.fontSize=104] - 比例尺数值字体字号
   * @param {string} [paramObject.fontColor='#221815'] - 比例尺数值字体颜色
   */
  createScale(scale, {
    left = 4000,
    top = 13812,
    fontSize = 104,
    fontColor = '#221815'
  } = {}) {
    if (!scale) throw new RangeError('比例尺参数错误');
    const {
      ctx
    } = this;
    ctx.fillText(`比例尺 1:${this.percentSign(scale)}`, left, top);
    ctx.font = `${fontSize}px "hanyizhongyuan"`;
    ctx.fillStyle = fontColor;
    const [scaleBarLeft, scaleBarTop, scaleBarSize] = [left + 971, top + 57, fontSize / 2];
    this.createScaleBar(scale, scaleBarLeft, scaleBarTop, scaleBarSize, fontColor);
  }
  /**
   * 比例尺图示
   * @param {number} scale 比例尺数值
   * @param {object} [paramObject={}] - 比例尺图示定位参数
   * @param {number} [paramObject.left=4971] - 比例尺图示相对左边距
   * @param {number} [paramObject.top=13869] - 比例尺图示相对上边距
   * @param {number} [paramObject.fontSize=52] - 比例尺图示字体字号
   * @param {string} [paramObject.fontColor='#221815'] - 比例尺图示字体颜色
   */
  createScaleBar(
    scale,
    scaleBarLeft,
    scaleBarTop,
    scaleBarSize,
    fontColor
  ) {
    const {
      ctx
    } = this;
    let [left, top, fontSize] = [scaleBarLeft, scaleBarTop, scaleBarSize];
    // 边框矩形 & 实体矩形
    ctx.strokeStyle = fontColor;
    ctx.fillStyle = fontColor;
    ctx.fillRect(left + 0 * 118, top, 118, 35);
    ctx.strokeRect(left + 1 * 118, top, 118, 35);
    ctx.fillRect(left + 2 * 118, top, 118, 35);
    ctx.strokeRect(left + 3 * 118, top, 118, 35);
    ctx.fillRect(left + 4 * 118, top, 236, 35);
    ctx.strokeRect(left + 6 * 118, top, 236, 35);
    ctx.fillRect(left + 8 * 118, top, 236, 35);
    // 3px线条border
    ctx.lineWidth = 3;
    ctx.moveTo(left, top);
    ctx.lineTo(left + 1180, top);
    ctx.moveTo(left, top + 35);
    ctx.lineTo(left + 1180, top + 35);
    ctx.stroke();
    // 图示单位
    const [tsLeft, tsTop] = [left + 1197, top - 11];
    let tsDw = '';
    // 图视数值
    let cmToM;
    if (scale < 50000) {
      cmToM = Math.round(scale / 100);
      tsDw = '米';
    } else {
      cmToM = scale / 100000;
      tsDw = '千米';
    }
    ctx.font = `${fontSize}px hanyizhongyuan`;
    ctx.fillStyle = fontColor;
    ctx.fillText(tsDw, tsLeft, tsTop);
    for (let i = 0; i < 11; i++) {
      const item = `${cmToM * i}`;
      const [itemLeft, itemTop] = [left + 118 * i - item.length * fontSize / 5, top - 78];
      if (i < 5) {
        ctx.fillText(item, itemLeft, itemTop);
      } else {
        if (i % 2 === 0) {
          ctx.fillText(item, itemLeft, itemTop);
        }
      }
    }
    console.log('比例尺√');
  }

  /**
   * 承编单位
   * @param {string} [cbdwName='国信司南（北京）地理信息技术有限公司'] - 承编单位字体名称
   * @param {object} [paramObject={}] - 承编单位定位参数
   * @param {number} [paramObject.left=6816] - 承编单位字体相对左边距
   * @param {number} [paramObject.top=13799] - 承编单位字体相对上边距
   * @param {number} [paramObject.fontSize=104] - 承编单位字体字号
   * @param {string} [paramObject.fontColor='#221815'] - 承编单位字体颜色
   */
  createCbdw(cbdwName = '国信司南（北京）地理信息技术有限公司', {
    left = 6816,
    top = 13799,
    fontSize = 104,
    fontColor = '#221815'
  } = {}) {
    const name = `承编单位：${cbdwName}`;
    const {
      ctx
    } = this;
    ctx.font = `${fontSize}px hanyizhongyuan`;
    ctx.fillStyle = fontColor;
    ctx.fillText(name, left, top);
    console.log('承编单位√');
  }
}
module.exports = A0Vertical;